l3kernel/l3regex.dtx

```diff
@@ -3359,19 +3384,37 @@
         \token_if_eq_meaning:NNTF #1 \@@_compile_special:N
           {
             \exp_after:wN \token_if_eq_charcode:NNTF \c_right_brace_str #2
-              { \if_false: { \fi: } \@@_compile_u_end: }
+              { \if_false: { \fi: } \l_@@_internal_b_tl }
               { #2 \@@_compile_u_loop:NN }
           }
           {
             \if_false: { \fi: }
             \__kernel_msg_error:nnx { regex } { u-missing-rbrace } {#2}
-            \@@_compile_u_end:
+            \l_@@_internal_b_tl
             #1 #2
           }
       }
   }
 %    \end{macrocode}
 % \end{macro}
+%
+% \begin{macro}{\@@_compile_ur_end:}
+%   For the |\ur{...}| construction, once we have extracted the
+%   variable's name, we simply insert the compiled regex (which the
+%   variable should be) into a non-capturing group (with no repetition)
+%   to respect the structure of regexes.  It might be possible to omit
+%   this group perhaps.
```

Comment:
Results may differ if the group is or is not there, for example (if I did the non-group version correctly, that is :)
```latex
\ExplSyntaxOn
\cs_gset_protected:Npn \__regex_compile_ur_end:
  {
    \tl_build_put_right:Nx \l__regex_build_tl
      {
        \exp_not:N \__regex_group_no_capture:nnnN
          { \exp_not:v { \l__regex_internal_a_tl } }
          { 1 } { 0 } \exp_not:N \c_false_bool
      }
  }

\tl_set:Nn \l_tmpa_tl { bc }
\regex_set:Nn \l_tmpa_regex { a|b }
\regex_show:n { ^\ur{l_tmpa_regex}c }
\regex_replace_all:nnN { ^\ur{l_tmpa_regex}c } { [\0] } \l_tmpa_tl
\tl_show:N \l_tmpa_tl

\cs_gset_protected:Npn \__regex_compile_ur_end:
  {
    \tl_build_put_right:Nx \l__regex_build_tl
      { \exp_not:v { \l__regex_internal_a_tl } }
  }

\tl_set:Nn \l_tmpa_tl { bc }
\regex_set:Nn \l_tmpa_regex { a|b }
\regex_show:n { ^\ur{l_tmpa_regex}c }
\regex_replace_all:nnN { ^\ur{l_tmpa_regex}c } { [\0] } \l_tmpa_tl
\tl_show:N \l_tmpa_tl
\stop
```

In the case with the group it matches `bc`, as would the regex if written in full as `^a|bc` (which seems to be correct):
```
> Compiled regex {^\ur {l_tmpa_regex}c}:
+-branch
  assertion: anchor at start (\A)
  ,-group begin (no capture)
  | char code 97 (a)
  +-branch
  | char code 98 (b)
  `-group end
  char code 99 (c).
<recently read> }
                 
l.67 \regex_show:n { ^\ur{l_tmpa_regex}c }
                                          
> \l_tmpa_tl=[bc].
<recently read> }
```

In the case without the group the `\regex_show:n` shows more or less what `\regex_show:n { ^a|bc }` would show, but it only matches `b` which seems wrong (I think?):
```latex
> Compiled regex {^\ur {l_tmpa_regex}c}:
+-branch
  assertion: anchor at start (\A)
+-branch
  char code 97 (a)
+-branch
  char code 98 (b)
  char code 99 (c).
<recently read> }
                 
l.79 \regex_show:n { ^\ur{l_tmpa_regex}c }
                                          
> \l_tmpa_tl=[b]c.
<recently read> }
```

Reply:
1. Actually the regex `^\ur{l_tmpa_regex}c` is equivalent to `^(a|b)c` but with a non-capturing group.  This is pretty different from `^a|bc`, which is equivalent to `(^a)|(bc)` (but without capturing of course).  Here you didn't see the difference because the `bc` branch ends up being the same.

2. I think it would be confusing if `^ \ur{l_tmpa_regex} c` could match something that does not contain `c`, as would happen for the `^a|bc` interpretation.  So I think it is best to interpret and document `\ur{...}` as introducing a non-capturing group.  My point there was an implementation note: for some simple regexes it is not necessary to introduce the overhead of an actual group.  To be precise, the overhead can be avoided if the inserted regex has a single branch at top level (no "`|`" alternative).  For instance inserting `A(B|C)D` inside some other regex can be done cheaper than it currently is.

3.  The oddity you found when constructing the regex by hand is because the regex is not a valid regex.  (This motivated me just now to log issue #884: `\regex_show:N` should ideally catch that problem.)

Reply:
Ah, so my second regex was indeed invalid!  That explains the difference then. Thanks!

Reply:
I should add that if someone wants to get "`^`(something)`c`" to turn into `^a|bc`, then the easiest way is to do

    \tl_set:Nn \l_myreg_tl { a|b }
    \cs_generate_variant:Nn \regex_set:Nn { Nx }
    \regex_set:Nx \l_my_regex { ^ \l_myreg_tl c }

namely to use TeX's standard expansion machinery.  I'll document this momentarily.